### DISCUSSION TITLE
[RFC] Actions 操作列表

### DISCUSSION THREAD

**Discussion by ONLY-yours** [ant-design/x#77] [RFC]
## Summary

新增名为 ActionBar - 操作列表 的新组件。目的快速配置一些 AI 场景下所需要的操作按钮/功能

## Motivation

在 AI 场景下有些通用、常见的操作列表，例如对话的时候会有一些「重新生成」、「复制」、「点赞」、「点踩」等操作
![image](https://github.com/user-attachments/assets/8ea113ac-1fae-4eb4-aa41-af626ebf9794)
当选项较多时，还需要提供操作收纳的能力
![image](https://github.com/user-attachments/assets/303246e8-a278-4203-bae2-cccfc6473432)
因此需要有个「操作列表」来帮助用户快速配置对应的能力

有些按钮定制化程度很高，但是有些操作又是通用的，我们会提供一些方式将一些通用的按钮能力提供出来，并支持用户自定义

### 组件名称

+ **英文名称:** Actions
+ **中文名称:** 操作列表

### API 设计

```typescript
interface ActionsProps {
	items: ActionItem[]
	rootClassName?: string;
	style?: React.CSSProperties
	variant?: 'borderless' | 'border'
}
interface ItemType {
	label?: string; // 自定义操作的显示标签
	key: string; // 自定义操作的唯一标识
	icon: ReactNode;// 自定义操作的图标
	label?: ReactNode; // 自定义操作的显示标签	
	onClick?: () => void; // 点击自定义操作按钮时的回调函数
}
interface SubItemType {
	key: string; // 自定义操作的唯一标识
	label?: string; // 自定义操作的显示标签
	icon?: ReactNode;// 自定义操作的图标
	children?: ActionBarItemType[]
	onClick?: () => void; // 点击自定义操作按钮时的回调函数
}
type ActionItemType =  ItemType|SubItemType
```

### 基本使用

![image](https://github.com/user-attachments/assets/44ee7cda-7e5a-4d3a-bb6d-7ef9efcfbc4c)

```tsx
import React, { useState } from 'react';
import { Actions } from '@ant-design/x';
import { RedoOutlined, CopyOutlined } from '@ant-design/icons'

const actionItems  = [
    {
      key: 'retry',
      icon: <RedoOutlined />,
    }, {
	key: 'copy',
	icon: <CopyOutlined />,
    }
  ]

const App: React.FC = () => {

  return (
    <Actions items={actionItems} />
  );
};

export default App;
```

### 配置 子菜单

![image](https://github.com/user-attachments/assets/303246e8-a278-4203-bae2-cccfc6473432)

```tsx
import React, { useState } from 'react';
import { Actions } from '@ant-design/x';
import { RedoOutlined, CopyOutlined, DeleteOutlined, ShareAltOutlined } from '@ant-design/icons'

const actionItems  = [
    {
      key: 'retry',
      icon: <RedoOutlined />,
    }, {
	key: 'copy',
	icon: <CopyOutlined />,
    }, {
	key: 'more',
	// icon:  <EllipsisOutlined />, // 不传使用默认的icon
	chidren: [
		{ key: 'share', label: '分享' , icon: <ShareAltOutlined />},
		{ key: 'import', label: '引用' },
          	{ key: 'delete', label: '删除' , icon: <DeleteOutlined />, onClick: () => {}},
	]
    }
  ]

const App: React.FC = () => {

  return (
    <Actions items={actionItems} />
  );
};

export default App;
```

### 配置变体

通过`variant`设置变体
![image](https://github.com/user-attachments/assets/f6f20ec3-bf80-4a9c-beb2-d2f98ce30fb5)

```tsx
import React, { useState } from 'react';
import { ActionBar } from '@ant-design/x';
import { RedoOutlined, CopyOutlined } from '@ant-design/icons'

const actionItems  = [
    {
      key: 'retry',
      icon: <RedoOutlined />,
    }, {
	key: 'copy',
	icon: <CopyOutlined />,
    }
  ]

const App: React.FC = () => {

  return (
    <Actions items={actionItems}  variant="border"/>
  );
};

export default App;
```

**Comment by zombieJ** [comment]
* antd 的命名规范一般叫 `items`
* and 看起来类似 Menu，似乎 API 直接对着 Menu 就行了？
* 然后 block 作为整行样式，variant 作为变体（有边框、无边框、透明等等）

**Reply by ONLY-yours** [reply]
我觉得和 Menu 还是不太一样的，我更倾向于是 popover 或者 dropdown 的嵌套情况
- 有些时候 弹出框会有一些类似 placement 之类的 api

**Comment by afc163** [comment]
ActionBar => Actions 如何？

在桌面端可能是一个 bar，但是在移动端可能是长按菜单的形式。是不是没必要在名字上做限定。

<img width="403" alt="image" src="https://github.com/user-attachments/assets/bfba0f7c-5a47-4625-9fac-2b1f857adba1">

![image](https://github.com/user-attachments/assets/e2dd69b8-8f35-4b29-b972-eece5925a66e)

![image](https://github.com/user-attachments/assets/323a09f1-25e3-4064-a0ed-dac4877d3e85)

**Reply by disturbwe** [reply]
Actions确实更响亮一点

**Comment by wzxinchen1** [comment]
对应的PR还没合并，https://github.com/ant-design/x/pull/238

**Comment by kimteayon** [comment]
这个还有人跟进吗？没人的话就我来跟进？
